### PR TITLE
Enable vectorization of libpgmath intrinsic calls

### PR DESCRIPTION
**Pull request by vjayathirtha-nv**
This PR adds PGMATH as one of the values for `-fveclib` flag. 
Also makes `-fveclib=PGMATH` the default option when 
a. User has not explicitly provided the `-fveclib` option.
b. The optimization level supports vectorization.

### CODE REVIEW

**Comment by vjayathirtha-nv** [comment]
Associated with [this PR](https://github.com/flang-compiler/llvm/pull/43)

File: lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -4351,6 +4349,12 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   if (Args.hasFlag(options::OPT_fvectorize, VectorizeAliasOption,
                    options::OPT_fno_vectorize, EnableVec))
     CmdArgs.push_back("-vectorize-loops");
+  
+  if (Args.getLastArg(options::OPT_fveclib))
+    Args.AddLastArg(CmdArgs, options::OPT_fveclib);
+  else if (Args.hasFlag(options::OPT_fvectorize, VectorizeAliasOption,
+                        options::OPT_fno_vectorize, EnableVec))
+    CmdArgs.push_back("-fveclib=PGMATH");
```
Comment (sscalpone) [inline lib/Driver/ToolChains/Clang.cpp]: What happens if someone is using a version of LLVM without support for -fveclib=PGMATH?

Comment (vjayathirtha-nv) [reply lib/Driver/ToolChains/Clang.cpp]: LLVM generates an error as it would with any option which it does not recognize.

Comment (sscalpone) [reply lib/Driver/ToolChains/Clang.cpp]: While flang has its own fork of llvm, flang still works with upstream llvm as well.  The changes for DWARF, for example, are enabled with a compile-time #ifdef in flang-compiler/flang.  Likely we need a similar mechanism here until we know that all of the upstream guides to building flang (not just blogs, but spack etc) are updated to refer to flang-compiler/flang.

Comment (schweitzpgi) [reply lib/Driver/ToolChains/Clang.cpp]: There is already a CMAKE variable that flang builders can use to include Flang-specific LLVM extensions. (FLANG_LLVM_EXTENSIONS)

File: lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -3286,8 +3286,6 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   else
     CmdArgs.push_back(Args.MakeArgString(getToolChain().getThreadModel()));
 
-  Args.AddLastArg(CmdArgs, options::OPT_fveclib);
-
```
Comment (schweitzpgi) [inline lib/Driver/ToolChains/Clang.cpp]: Why is this moved 1000 lines down?

File: lib/Driver/ToolChains/Clang.cpp (hunk 1 above)
Comment (schweitzpgi) [reply lib/Driver/ToolChains/Clang.cpp]: I don't agree that -fvectorize should imply extra hidden arguments. The user should be in control of the kind of vectorization and the library calls to be used and these should remain independent.

Comment (vjayathirtha-nv) [reply lib/Driver/ToolChains/Clang.cpp]: I thought we wanted `-fveclib=PGMATH` to be the default. I can undo this.

**Comment by vjayathirtha-nv** [comment]
I have removed `-fveclib=PGMATH` as the default flag.

File: lib/Driver/ToolChains/Clang.cpp (hunk 1 above)
Comment (sscalpone) [reply lib/Driver/ToolChains/Clang.cpp]: We do want PGMATH to be the default but the flang ecosystem is not ready for that b/c we still want to build with llvm from llvm.org or pre-installed llvm in a distro; we need to move these changes upstream.  Unlike SVML, libpgmath is always available with flang so making it the default will not have any negative user impact.

Comment (sscalpone) [reply lib/Driver/ToolChains/Clang.cpp]: Please add FLANG_LLVM_EXTENSIONS to flang-driver in a subsequent commit; soon please.

File: lib/Driver/ToolChains/Clang.cpp (hunk 2 above)
Comment (vjayathirtha-nv) [reply lib/Driver/ToolChains/Clang.cpp]: To reuse the `EnableVec` and `VectorizeAliasOption` variables in line 4356.

Comment (schweitzpgi) [reply lib/Driver/ToolChains/Clang.cpp]: Why not just put the -fveclib change here. Why do we care about the vectorizer command line being present? It's default on all the time, right?

+#ifdef FLANG_LLVM_EXTENSIONS
+  if(Args.getLastArg(options::OPT_fveclib))
+    Args.AddLastArg(CmdArgs, options::OPT_fveclib);
+  else 
+    CmdArgs.push_back("-fveclib=PGMATH");
+#else
+  Args.AddLastArg(CmdArgs, options::OPT_fveclib);
+#endif

Comment (vjayathirtha-nv) [reply lib/Driver/ToolChains/Clang.cpp]: It doesn't make sense to add the `-fveclib=PGMATH` flag when there is no vectorization taking place.

**Comment by schweitzpgi** [comment]
+1
